zrml/asset-router/src/lib.rs

```diff
@@ -162,645 +183,252 @@ pub mod pallet {
-                return Err(Error::<T>::AmountIntoBalanceFailed.into());
-            } else {
-                by_amount.abs()
-            };
-
-            let by_balance = TryInto::<Self::Balance>::try_into(by_amount_abs)
-                .map_err(|_| Error::<T>::AmountIntoBalanceFailed)?;
-            if by_amount.is_positive() {
-                Self::deposit(currency_id, who, by_balance)
-            } else {
-                Self::withdraw(currency_id, who, by_balance).map(|_| ())
-            }
-        }
-    }
-
-    impl<T: Config> MultiLockableCurrency<T::AccountId> for Pallet<T> {
-        type Moment = T::BlockNumber;
-
-        fn set_lock(
-            lock_id: LockIdentifier,
-            currency_id: Self::CurrencyId,
-            who: &T::AccountId,
-            amount: Self::Balance,
-        ) -> DispatchResult {
-            if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
-                return <T::Currencies as MultiLockableCurrency<T::AccountId>>::set_lock(
-                    lock_id, currency, who, amount,
-                );
-            }
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
 
-            Err(Error::<T>::Unsupported.into())
-        }
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+        fn on_idle(_: T::BlockNumber, mut remaining_weight: Weight) -> Weight {
+            let mut destroy_assets = DestroyAssets::<T>::get();
 
-        fn extend_lock(
-            lock_id: LockIdentifier,
-            currency_id: Self::CurrencyId,
-            who: &T::AccountId,
-            amount: Self::Balance,
-        ) -> DispatchResult {
-            if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
-                return <T::Currencies as MultiLockableCurrency<T::AccountId>>::extend_lock(
-                    lock_id, currency, who, amount,
-                );
+            if destroy_assets.len() == 0 {
+                return remaining_weight.saturating_sub(T::DbWeight::get().reads(1));
             }
 
-            Err(Error::<T>::Unsupported.into())
-        }
-
-        fn remove_lock(
-            lock_id: LockIdentifier,
-            currency_id: Self::CurrencyId,
-            who: &T::AccountId,
-        ) -> DispatchResult {
-            if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
-                return <T::Currencies as MultiLockableCurrency<T::AccountId>>::remove_lock(
-                    lock_id, currency, who,
-                );
+            remaining_weight =
+                remaining_weight.saturating_sub(T::DbWeight::get().reads_writes(1, 1));
+
+            'outer: while let Some(mut asset) = destroy_assets.pop() {
+                let mut safety_counter: u8 = 0;
+
+                while (*asset.state() != DestructionState::Destroyed
+                    || *asset.state() != DestructionState::Indestructible)
+                    && safety_counter < 6
```

Comment:
https://github.com/zeitgeistpm/zeitgeist/pull/1201/commits/965e26e5f96811b51ecd0ddf8280bf29eddad98c https://github.com/zeitgeistpm/zeitgeist/pull/1201/commits/64d84e2d0d9f52ddc74300384ea9829df59d22be

In regards to catching when unreachable code was triggered, this might be the solution: https://github.com/zeitgeistpm/zeitgeist/blob/64d84e2d0d9f52ddc74300384ea9829df59d22be/zrml/asset-router/src/lib.rs#L278-L282
I think we should use this approach in every instance where unreachable code is triggered. If you agree, I'll create a PR to add this to the style guide.
This approach seems verify powerful in conjunction with prop and fuzz testing (either of them).

Reply:
Is there any non-academic difference between `debug_assert` and `#[cfg(test)] assert!`?

Reply:
Why exactly do you mean by `non-academic`?

The difference is that `debug_assert` will crash clients build in debug mode, while `#[cfg(test)] assert!` won't.

Reply:
"Non-academic" refers to the difference between the two in the intermediate format, which is a thing (I think).

Is there any reason why someone would build a client in debug mode and then get angry at us if a legit logic error caused a crash? But yeah, I guess `#[cfg(test)] assert!` expresses exactly what we want, so that's fine with me. Might be a good idea to just introduce an `unreachable!` macro?

Reply:
I think that would be strange behavior by the user, but on the other side it's not documented that debug clients might behave differently than production clients. IRC that is also the main argument from SRLabs; That one group of nodes would crash and another wouldn't, thus leading to diverging states between clients with the consequence of never being able to find consensus about a common state again.
A macro makes sense from my perspective, assuming it can cover all different cases.

Reply:
Unreachable macro: https://github.com/zeitgeistpm/zeitgeist/commit/979655cb4c3d047cf89e0b48466de498a68a64f0